src/services/sablier.js

```diff
@@ -0,0 +1,323 @@
+    : EthersBigNumber.from(0);
+  return approvedAmountBN;
+};
+
+export const getSmartWalletTxFee = async (transaction: Object, useGasToken: boolean): Promise<Object> => {
+  const defaultResponse = { fee: new BigNumber('0'), error: true };
+  const estimateTransaction = {
+    data: transaction.data,
+    recipient: transaction.to,
+    value: transaction.amount,
+  };
+
+  const estimated = await smartWalletService
+    .estimateAccountTransaction(estimateTransaction)
+    .then(result => buildTxFeeInfo(result, useGasToken))
+    .catch((e) => {
+      reportLog('Error getting sablier fee for transaction', {
+        ...transaction,
+        message: e.message,
+      }, Sentry.Severity.Error);
+      return null;
+    });
+
+  if (!estimated) {
+    return defaultResponse;
+  }
+
+  return estimated;
+};
+
+const getTxFeeAndTransactionPayload = async (_transactionPayload, useGasToken): Promise<Object> => {
+  const { fee: txFeeInWei, gasToken, error } = await getSmartWalletTxFee(_transactionPayload, useGasToken);
+  let transactionPayload = _transactionPayload;
+  if (gasToken) {
+    transactionPayload = { ...transactionPayload, gasToken };
+  }
+
+  transactionPayload = { ...transactionPayload, txFeeInWei };
+
+  if (error) {
+    return null;
+  }
+
+  return {
+    gasToken,
+    txFeeInWei,
+    transactionPayload,
+  };
+};
+
+export const getApproveFeeAndTransaction = async (assetData: Asset, useGasToken: boolean): Promise<Object> => {
+  const { symbol, decimals, address: tokenAddress } = assetData;
+  const rawValue = 1000000000;
+  const valueToApprove = utils.parseUnits(rawValue.toString(), decimals);
+
+  const data = encodeContractMethod(ERC20_CONTRACT_ABI, 'approve', [SABLIER_CONTRACT_ADDRESS, valueToApprove]);
+
+  const transactionPayload = {
+    amount: 0,
+    to: tokenAddress,
+    symbol,
+    contractAddress: tokenAddress,
+    decimals,
+    data,
+    extra: {
+      sablierApproval: {
+        symbol,
+      },
+    },
+    tag: SABLIER_ALLOWANCE,
+  };
+
+  return getTxFeeAndTransactionPayload(transactionPayload, useGasToken);
+};
+
+export const getCancellationFeeAndTransaction = async (stream: Stream, useGasToken: boolean): Promise<Object> => {
+  const transactionData = encodeContractMethod(SABLIER_ABI, 'cancelStream', [
+    stream.id,
+  ]);
+  const transactionPayload = {
+    to: SABLIER_CONTRACT_ADDRESS,
+    data: transactionData,
+    amount: 0,
+    symbol: ETH,
+    tag: SABLIER_CANCEL_STREAM,
+    extra: {
+      assetAddress: stream.token.id,
+      amount: stream.deposit,
+      contactAddress: stream.recipient,
+      streamId: stream.id,
+    },
+  };
+
+  return getTxFeeAndTransactionPayload(transactionPayload, useGasToken);
+};
+
+export const getCreateStreamFeeAndTransaction = async (
+  sender: string,
+  receiver: string,
+  amount: EthersBigNumber,
+  tokenAddress: string,
+  startTimestamp: number,
+  endTimestamp: number,
+  useGasToken: boolean,
+): Promise<Object> => {
```

Comment:
Oh i've just seen this - but does this need to be a promise? Maybe error control? Please let me know if i've misunderstood something here.

Reply:
I've removed `async`s but the promise is correct: `getTxFeeAndTransactionPayload` is `async` so it returns the promise and so these functions also return promises

Reply:
Got it - thank you.